### PR TITLE
[Test] Reset timer for every unit test and provide timestamps for log messages

### PR DESCRIPTION
**Pull request by KarlK90**
## Description

Previously the unit-tests where run with a continuously running timer, which can trigger timing bugs in unrelated unit tests. To remedy this we now reset the timer for each unit test. All events in log messages now also contain a timestamp and key events show how long they have been held/released. This is meant as a convenience feature for debugging and developing unit-tests.

e.g. the quite complex `ANewTapWithinTappingTermIsBuggy` test on failure before:

```
Tapping.ANewTapWithinTappingTermIsBuggy failed!
[ TRACE    ] Key pressed:  (7,0)
[ TRACE    ] Key released: (7,0)
[ TRACE    ] Keyboard Report: Mods (0) Keys (19)
[ TRACE    ] Keyboard Report: Mods (0) Keys ()
[ TRACE    ] Key pressed:  (7,0)
[ TRACE    ] Keyboard Report: Mods (0) Keys (19)
[ TRACE    ] Key released: (7,0)
[ TRACE    ] Keyboard Report: Mods (0) Keys ()
[ TRACE    ] Key pressed:  (7,0)
[ TRACE    ] Key released: (7,0)
[ TRACE    ] Keyboard Report: Mods (0) Keys (19)
[ TRACE    ] Keyboard Report: Mods (0) Keys ()
[ TRACE    ] Key pressed:  (7,0)
[ TRACE    ] Key released: (7,0)
[ TRACE    ] Keyboard Report: Mods (2) Keys ()
[ TRACE    ] Keyboard Report: Mods (0) Keys ()
[ TRACE    ] Keyboard Report: Mods (0) Keys (19)
[ TRACE    ] Keyboard Report: Mods (0) Keys ()
[ INFO     ] TestFixture clean-up start.
[ INFO     ] TestFixture clean-up end.
```

and with these changes:

```
Tapping.ANewTapWithinTappingTermIsBuggy failed!
[ LEVEL    ] [TIME] [EVENT]
[ INFO     ] 0      tapping term is 200ms
[ TRACE    ] 0      pressed:  MT(MOD_LSFT, KC_P)
[ TRACE    ] 0      1 keyboard task loop
[ TRACE    ] 1      released: MT(MOD_LSFT, KC_P) was pressed for 1ms
[ TRACE    ] 1      1 keyboard task loop
[ TRACE    ] 1      report:   (KC_P) []
[ TRACE    ] 1      report:   empty
[ TRACE    ] 2      pressed:  MT(MOD_LSFT, KC_P)
[ TRACE    ] 2      1 keyboard task loop
[ TRACE    ] 2      report:   (KC_P) []
[ TRACE    ] 3      released: MT(MOD_LSFT, KC_P) was pressed for 1ms
[ TRACE    ] 3      201 keyboard task loops
[ TRACE    ] 3      report:   empty
[ TRACE    ] 204    pressed:  MT(MOD_LSFT, KC_P)
[ TRACE    ] 204    1 keyboard task loop
[ TRACE    ] 205    released: MT(MOD_LSFT, KC_P) was pressed for 1ms
[ TRACE    ] 205    201 keyboard task loops
[ TRACE    ] 205    report:   (KC_P) []
[ TRACE    ] 205    report:   empty
[ TRACE    ] 406    pressed:  MT(MOD_LSFT, KC_P)
[ TRACE    ] 406    200 keyboard task loops
[ TRACE    ] 606    released: MT(MOD_LSFT, KC_P) was pressed for 200ms
[ TRACE    ] 606    1 keyboard task loop
[ TRACE    ] 606    report:   () [KC_LEFT_SHIFT]
[ TRACE    ] 606    report:   empty
[ INFO     ] 607    test fixture clean-up start.
[ TRACE    ] 607    2000 keyboard task loops
[ TRACE    ] 2607   2000 keyboard task loops
[ INFO     ] 4607   test fixture clean-up end.
```

## Types of Changes

- [x] Core
- [ ] Bugfix
- [x] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* Cherry-picked from #15847 which will be WIP for longer time.

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by KarlK90** [comment]
CC: @filterpaper @precondition @joukewitteveen as you are involved with unit tests recently and might this feature useful?

**Comment by precondition** [comment]
I'd add an explanation of what the numbers after [TRACE] mean in the docs page on unit tests. It's not obvious that it is a timestamp.

File: tests/test_common/test_keymap_key.cpp
Comment (joukewitteveen) [inline tests/test_common/test_keymap_key.cpp:27]:
```suggestion
                                                           << "that was already pressed! Check the test code." << std::endl;
```
Decoupling the tests sounds like a good idea to me, but I don't think it will impact the tapdance tests that I am trying to write.

**Comment by KarlK90** [comment]
@precondition I did go a bit further and included a header for the output that explains each column.

For better introspection I prototyped a mechanism that resolves the usb report keys and mods to readable constants, furthermore the QMK keycode for each button is now stored in the keycode on construction (although with a bit ugly C stringify magic). I updated the log output in the description of this PR accordingly.

File: tests/test_common/test_keymap_key.cpp
Comment (KarlK90) [reply tests/test_common/test_keymap_key.cpp]: What would help you developing the tap dance unit-tests?

File: tests/test_common/test_keymap_key.cpp
Comment (joukewitteveen) [inline tests/test_common/test_keymap_key.cpp:27]:
```suggestion
    EXPECT_FALSE(matrix_is_on(position.row, position.col)) << "Tried to press key " << this->name << " that was already pressed! Check the test code." << std::endl;
```

File: tests/test_common/test_keymap_key.cpp
Comment (joukewitteveen) [inline tests/test_common/test_keymap_key.cpp:35]:
```suggestion
    EXPECT_TRUE(matrix_is_on(this->position.row, this->position.col)) << "Tried to release key " << this->name << " that wasn't pressed before! Check the test code." << std::endl;
```

File: tests/test_common/test_keymap_key.cpp
Comment (joukewitteveen) [reply tests/test_common/test_keymap_key.cpp]: Some more time; I'm making progress and already found a few bugs.

Comment (KarlK90) [reply tests/test_common/test_keymap_key.cpp]: Alright, I'm not trying to push anything. Just asking if you perhaps need more facilities that would help you.

File: tests/test_common/test_keymap_key.cpp
Comment (KarlK90) [reply tests/test_common/test_keymap_key.cpp]: Thanks, I accidentally force pushed over your suggestions.

**Comment by joukewitteveen** [comment]
> CC: @filterpaper @precondition @joukewitteveen as you are involved with unit tests recently and might this feature useful?

I did not test the code in this PR and the changes in #16394 would be affected by the changes in this PR: the `KEYMAP_KEY` macro is not yet used in #16394.

In case of failing tests related to tap dance, I can see how it can be pleasant to restart the timer for each test. That would keep the numbers down a bit.

**Review by drashna — APPROVED** [review]
`__attribute__((weak))` ✅

**Review by joukewitteveen — CHANGES_REQUESTED** [review]
The tap dance tests added with #16394 should also be updated to use the `KEYMAP_KEY` macro.

**Comment by joukewitteveen** [comment]
> The tap dance tests added with #16394 should also be updated to use the `KEYMAP_KEY` macro.

Note that these updates of the tap dance tests are still missing from this PR.

**Comment by KarlK90** [comment]
> > The tap dance tests added with #16394 should also be updated to use the `KEYMAP_KEY` macro.
> 
> Note that these updates of the tap dance tests are still missing from this PR.

Thanks for the reminder, though the `KEYMAP_KEY` macro no longer exists and was replaced with a reverse lookup mechanism that constructs a identifier from a keycode.

**Review by joukewitteveen — APPROVED** [review]
> [...] the KEYMAP_KEY macro no longer exists and was replaced with a reverse lookup mechanism that constructs a identifier from a keycode.

Cool, sorry for the noise.

**Comment by KarlK90** [comment]
On hold until DD keycodes have been merged

**Comment by sigprof** [comment]
But #18643 had been merged… or do you mean something else?

**Comment by KarlK90** [comment]
> But #18643 had been merged… or do you mean something else?

You are right, I missed that. I'll generate the keycode lookup table from the keycode specifications.

File: quantum/quantum_keycodes.h
Comment (zvecr) [inline quantum/quantum_keycodes.h:107]: None of these `IS_*` should be introduced as they already exist <https://github.com/qmk/qmk_firmware/blob/master/quantum/keycodes.h#L1276-L1323>

Comment (KarlK90) [reply quantum/quantum_keycodes.h:107]: Sure, they are a leftover from before these where introduced with the DD keycodes.

Comment (KarlK90) [reply quantum/quantum_keycodes.h]: They are gone now.

**Review by tzarc — APPROVED** [review]

**Pull request merged**
